gravatar-quickeditor/src/main/java/com/gravatar/quickeditor/data/storage/DataStoreProfileStorage.kt

```diff
@@ -0,0 +1,35 @@
+package com.gravatar.quickeditor.data.storage
+
+import androidx.datastore.core.DataStore
+import androidx.datastore.preferences.core.Preferences
+import androidx.datastore.preferences.core.booleanPreferencesKey
+import androidx.datastore.preferences.core.edit
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.withContext
+import java.io.IOException
+
+internal class DataStoreProfileStorage(
+    private val dataStore: DataStore<Preferences>,
+    private val dispatcher: CoroutineDispatcher,
+) : ProfileStorage {
+    private val loginIntroShownKeyPrefix = "login_intro_shown"
+
+    override suspend fun setLoginIntroShown(emailHash: String): Unit = withContext(dispatcher) {
+        dataStore.edit { preferences ->
+            preferences[booleanPreferencesKey(emailHash.loginIntroShownKey)] = true
+        }
+    }
+
+    @Suppress("SwallowedException")
+    override suspend fun getLoginIntroShown(emailHast: String): Boolean = withContext(dispatcher) {
+        try {
+            dataStore.data.first()[booleanPreferencesKey(emailHast.loginIntroShownKey)] ?: false
+        } catch (exception: IOException) {
+            false
+        }
+    }
+
+    private val String.loginIntroShownKey: String
+        get() = "${loginIntroShownKeyPrefix}_$this"
```

Comment:
❓ We are adding the prefix just in case we want to store more info in this dataStore, right?

Reply:
Correct. I didn't want to use the token storage for that, but I also didn't want to create many files for potential flags in the future. I'd be happy to change that if you think otherwise.

Reply:
It makes sense to me, and it's not going to hurt. I'll keep it as it is.